library/src/main/java/com/timehop/stickyheadersrecyclerview/StickyRecyclerHeadersDecoration.java

```diff
@@ -5,212 +5,285 @@
-          if (orientation == LinearLayoutManager.VERTICAL &&
-              nextView.getTop() - secondHeader.getHeight() - firstHeader.getHeight() < 0) {
-            translationY += nextView.getTop() - secondHeader.getHeight() - firstHeader.getHeight();
-          } else if (orientation == LinearLayoutManager.HORIZONTAL &&
-              nextView.getLeft() - secondHeader.getWidth() - firstHeader.getWidth() < 0) {
-            translationX += nextView.getLeft() - secondHeader.getWidth() - firstHeader.getWidth();
-          }
+
+    private final StickyRecyclerHeadersAdapter mAdapter;
+    private final LongSparseArray<View> mHeaderViews = new LongSparseArray<>();
+    private final SparseArray<Rect> mHeaderRects = new SparseArray<>();
+    private final LongSparseArray<State> mHeaderViewStates = new LongSparseArray<>();
+    private final LongSparseArray<RecyclerView.ViewHolder> mHeaderViewHolders = new LongSparseArray<>();
+
+    public StickyRecyclerHeadersDecoration(StickyRecyclerHeadersAdapter adapter) {
+        mAdapter = adapter;
+    }
+
+    @Override
+    public void getItemOffsets(Rect outRect, View view, RecyclerView parent, RecyclerView.State state) {
+        super.getItemOffsets(outRect, view, parent, state);
+        int orientation = getOrientation(parent);
+        int itemPosition = parent.getChildPosition(view);
+        if (hasNewHeader(itemPosition)) {
+            View header = getHeaderView(parent, itemPosition);
+            if (orientation == LinearLayoutManager.VERTICAL) {
+                outRect.top = header.getHeight();
+            } else {
+                outRect.left = header.getWidth();
+            }
+        }
+    }
+
+    @Override
+    public void onDrawOver(Canvas canvas, RecyclerView parent, RecyclerView.State state) {
+        super.onDrawOver(canvas, parent, state);
+        int orientation = getOrientation(parent);
+        mHeaderRects.clear();
+
+        if (parent.getChildCount() > 0 && mAdapter.getItemCount() > 0) {
+            // draw the first visible child's header at the top of the view
+            View firstView = parent.getChildAt(0);
+            int firstPosition = parent.getChildPosition(firstView);
+            long firstHeaderId = mAdapter.getHeaderId(firstPosition);
+            if (firstHeaderId >= 0) {
+                View firstHeader = getHeaderView(parent, firstPosition);
+                View nextView = getNextView(parent);
+                int leftOffset = firstView.getLeft() - firstHeader.getWidth();
+                int topOffset = firstView.getTop() - firstHeader.getHeight();
+
+
+                if (orientation == LinearLayoutManager.VERTICAL) {
+                    if (topOffset == 0) {
+                        // The view is directly below the header but not under
+                        // make sure the header knows it's in inline mode
+                        setHeaderIsInline(firstHeaderId, firstPosition);
+                    } else if (topOffset < 0) {
```

Comment:
What about when topOffset > 0? No state would be assigned. If that isn't possible then just use an else statement and add a comment about it.